src/hidra/receiver/plugins/asapo_producer.py

```diff
@@ -86,244 +128,146 @@ class Ignored(Exception):
-#            self.data_type = "data"
-#        elif mode == "DEFAULT_INGEST_MODE":
-#            self.ingest_mode = asapo_producer.DEFAULT_INGEST_DATA
-#            self.data_type = "data"
-        else:
-            raise utils.NotSupported("Ingest mode '{}' is not supported"
-                                     .format(mode))
+        Args:
+            local_path: The absolute path where the file was written
+            metadata: The metadata to send as dict
+            data (optional): the data to send
+        """
 
-    def get_data_type(self):
-        return self.data_type
-
-    def _create_producer(self, stream):
-        token = self.token or self._get_token()
-        beamtime = self.beamtime or self._get_beamtime()
-        self.log.debug("type of beamtime%s", type(beamtime))
-
-        config = dict(
-            endpoint=self.endpoint,
-            beamtime_id=beamtime,
-            beamline=self.beamline,
-            stream=stream,
-            token=token,
-            nthreads=self.n_threads,
-            timeout_sec=self.timeout
-        )
+        if self._config_is_modified() or self.asapo_worker is None:
+            self.config.update(utils.load_config(self.config["user_config_path"]))
+            self.asapo_worker = AsapoWorker(self.config)
 
-        self.log.info("Create producer with config=%s", config)
+        self.asapo_worker.send_message(local_path, metadata)
 
-        self.stream_info[stream] = {
-            "producer": asapo_producer.create_producer(**config),
-        }
+    def _get_config_time(self, file_path):
+        start = time()
+        while True:
+            try:
+                return path.getmtime(file_path)
+            except OSError as err:
+                if time() - start > self.config_timeout:
+                    raise err
+                self.log.warn(
+                    "Retrying writing data due to error: {}".format(err))
+                sleep(1)
+
+    def _config_is_modified(self):
+        ts = time()
+        if (ts - self.check_time) > self.timeout:
+            file_path = self.config["user_config_path"]
+            config_time = self._get_config_time(file_path)
+            if self.config_time != config_time:
+                self.config_time = config_time
+                return True
+        self.check_time = ts
+        return False
 
-    def _get_token(self):
-        with open(self.token_file, "r") as f:
-            token = f.read().replace('\n', '')
+    def stop(self):
+        """ Clean up """
+        self.asapo_worker.stop()
 
-        return token
 
-    def _get_beamtime(self):
-        (_, _, filenames) = next(os.walk(self.beamtime_file_path))
-        self.log.debug("filesnames = %s", filenames)
-        for name in filenames:
-            search = re.search(self.beamtime_file_regex, name)
-            if search:
-                matched = search.groupdict()
+class AsapoWorker:
+    def __init__(self, config):
```

Comment:
Here I would use keyword arguments. The constructor can then be called as `AsapoWorker(**config)`.